### DISCUSSION TITLE
Is it possible/practical to run cibuildwheel locally?

### DISCUSSION THREAD

**Discussion by ndw** [pypa/cibuildwheel#1473] [General]
I have `cibuildwheel` running successfully in GitHub actions for Linux and Windows Intel wheels and Mac universal wheels. That's all working just fine.

I'd like to be able to make Linux ARM wheels (to, for example, run in containers on Apple Silicon Macs). There's a fair bit of complexity in the GitHub action because we don't actually have the sources for building the libraries that are being linked in the repository. The action grabs them from another server and builds on top of them. That works, but looking at the instructions for doing QEMU emulation to get the ARM wheels doesn't fill me with confidence that I'll be able to navigate *that* much more complexity.

Instructions, pointers, etc. for building wheels in, for example a local docker container, would be much appreciated.

**Comment by Czaki** [comment]
Building Linux wheels is safe. You will end with few additional docker images on machine

**Reply by ndw** [reply]
`cibuildwheel` falls at the interesection of several things I know only enough about to be dangerous. Any pointers to instructions for configuring local Docker wheel building would be most appreciated.

**Reply by Czaki** [reply]
You have ARM machine?

**Reply by ndw** [reply]
I have an Appile Silicon Mac running a Linux/ARM VM, I believe.

**Reply by Czaki** [reply]
Did you check plain `cibuildwheel --plaform=linux .` (where `.` is path to project)?

**Reply by ndw** [reply]
I'm not absolutely certain I installed `cibuildwheel` correctly in the container, but it falls over almost immediately:

```
[root@ec17c2c8cea8 eec_pypi]# /opt/_internal/cpython-3.8.16/bin/cibuildwheel --platform=linux .

     _ _       _ _   _       _           _
 ___|_| |_ _ _|_| |_| |_ _ _| |_ ___ ___| |
|  _| | . | | | | | . | | | |   | -_| -_| |
|___|_|___|___|_|_|___|_____|_|_|___|___|_|

cibuildwheel version 2.12.1

Build options:
  platform: linux
  architectures: x86_64
  build_selector:
    build_config: *
    skip_config: cp36* cp37* pp* cp*musl*
    requires_python: >=3.8
    prerelease_pythons: False
  container_engine: docker
  output_dir: /src/build/eec_pypi/wheelhouse
  package_dir: /src/build/eec_pypi
  test_selector:
    skip_config:
  before_all:
  before_build:
  before_test:
  build_frontend: pip
  build_verbosity: 0
  config_settings:
  dependency_constraints: pinned
  environment:
    LD_LIBRARY_PATH="/project/libs/nix"
  manylinux_images:
    x86_64: quay.io/pypa/manylinux2014_x86_64:2023-03-05-271004f
    i686: quay.io/pypa/manylinux2014_i686:2023-03-05-271004f
    pypy_x86_64: quay.io/pypa/manylinux2014_x86_64:2023-03-05-271004f
    aarch64: quay.io/pypa/manylinux2014_aarch64:2023-03-05-271004f
    ppc64le: quay.io/pypa/manylinux2014_ppc64le:2023-03-05-271004f
    s390x: quay.io/pypa/manylinux2014_s390x:2023-03-05-271004f
    pypy_aarch64: quay.io/pypa/manylinux2014_aarch64:2023-03-05-271004f
    pypy_i686: quay.io/pypa/manylinux2014_i686:2023-03-05-271004f
  musllinux_images:
    x86_64: quay.io/pypa/musllinux_1_1_x86_64:2023-03-05-271004f
    i686: quay.io/pypa/musllinux_1_1_i686:2023-03-05-271004f
    aarch64: quay.io/pypa/musllinux_1_1_aarch64:2023-03-05-271004f
    ppc64le: quay.io/pypa/musllinux_1_1_ppc64le:2023-03-05-271004f
    s390x: quay.io/pypa/musllinux_1_1_s390x:2023-03-05-271004f
  repair_command: auditwheel repair -w {dest_dir} {wheel}
  test_command:
  test_extras:
  test_requires:

Cache folder: /root/.cache/cibuildwheel

Here we go!

Traceback (most recent call last):
  File "/opt/_internal/cpython-3.8.16/bin/cibuildwheel", line 8, in <module>
    sys.exit(main())
  File "/opt/python/cp38-cp38/lib/python3.8/site-packages/cibuildwheel/__main__.py", line 141, in main
    build_in_directory(args)
  File "/opt/python/cp38-cp38/lib/python3.8/site-packages/cibuildwheel/__main__.py", line 293, in build_in_directory
    cibuildwheel.linux.build(options, tmp_path)
  File "/opt/python/cp38-cp38/lib/python3.8/site-packages/cibuildwheel/linux.py", line 379, in build
    subprocess.run(
  File "/opt/python/cp38-cp38/lib/python3.8/subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/opt/python/cp38-cp38/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/opt/python/cp38-cp38/lib/python3.8/subprocess.py", line 1704, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'docker'
```

**Reply by henryiii** [reply]
It shouldn't be running in the container, it runs a container for you (notice it's trying to call `docker`).

**Reply by ndw** [reply]
Okay. I did the build in a Docker Linux container, but I can build in a VMWare VM instead. And then I see that `cibuildwheel` is trying to start a docker container to do the job, so I think that's progress. It's trying to start an `x86_64` container not an `aarch64` container but I imagine that's a configuration error on my part.

**Reply by henryiii** [reply]
You don’t need a vm either. As long as you tell it `--platform linux` you can (and should) run it natively. You might need to pass the archs flag too, will have to check and see if we get the default right on AS.

**Reply by joerick** [reply]
You might be overthinking this...! Don't worry about VMs etc on Mac, Docker Desktop for Mac handles that for you. As long as you have Docker for Mac running, you should be able to run `cibuildwheel --platform linux` from a Mac terminal and it will start building inside a docker container. It will even try to emulate other archs if you want, but i've had mixed success with that on mac.

**Reply by ndw** [reply]
Ah. I may have failed to make clear that part of what needs to be in the wheel is a binary shared library. So I have to build that on a Linux platform of some sort.

**Reply by henryiii** [reply]
It does run in a Linux docker container - it always does this for Linux. `--platform linux` will always start up a Linux Docker image. On macOS, Docker sets up a light weight VM when you run a Linux Docker container.

If you need to add something to the container, you have to either build it in `CIBW_BEFORE_ALL`/`tool.cibuildwheel.before-all` or add it to the folder that gets mounted.

**Reply by ndw** [reply]
Right. I got it to work. Thank you for all your help!

**Comment by tabbyrobin** [comment]
Not sure if I should start a new conversation for this? I have a related question: Is running cibuildwheel locally a goal/feature of the project? Will it be supported indefinitely, with (reasonably) full feature parity to cloud runs?

I see that it is [mentioned](https://cibuildwheel.pypa.io/en/stable/setup/#local) in the docs, but there it is described as a convenient thing to do before doing the real (in-cloud) thing.

The context for my question is for enabling to verify reproducible builds. For this, the exact same logic needs to run in-cloud and locally.

**Reply by Czaki** [reply]
It depends what you mean locally. On your machine. I expect yes. With side effect on Windows and macOS of installing python from sources selected by project. That may not be suitable for all personal machines used for development purposes.

**Reply by henryiii** [reply]
An example is if you pre-install a python.org Python on your machine, cibuildwheel will use that instead of uninstalling yours and installing its own. It might be a different version or be native/universal when we would have selected the other one. That's exactly the same on CI, actually, but it's more likely to happen locally. And it's useful in CI in some cases, like when you need the one we don't use.

But yes, mostly the same, minus interactions with installed software on non-Linux systems.

**Reply by tabbyrobin** [reply]
Ok, great! Thanks
